compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -175,10 +175,26 @@ impl<'a> AstValidator<'a> {
                     }
                 }
             }
+            TyKind::AnonymousStruct(ref fields, ..) | TyKind::AnonymousUnion(ref fields, ..) => {
+                self.with_banned_assoc_ty_bound(|this| {
+                    walk_list!(this, visit_struct_field_def, fields)
+                });
+            }
             _ => visit::walk_ty(self, t),
         }
     }
 
+    fn visit_struct_field_def(&mut self, field: &'a FieldDef) {
+        self.check_anonymous_field(field);
```

Comment:
If `field.ident` is not an underscore here, then you can simply delegate to the default visitor (`self.visit_field_def()`) and it will do the right thing.
It will simplify the logic inside `fn check_anonymous_field`.

Reply:
I tried to implement this but I realized we would lose the `anonymous {} not allowed with a named identifier` message. Then I tried to move that logic to `deny_anonymous_field` but then an `Enum` with an unnamed struct would show that message even though an unnamed struct is not allowed in any `Enum` variant, so the only option is to remove that message.

I consider that error message important because C lang allows declaring things like:
```C
struct student{
    char name[30];
    struct avg {
        int sub1, sub2, sub3;
        float average;
    } avg1;
};
```
and knowing the objective of this RFC is to interface better with C code, in case the user duplicates the C struct declaration as it is I think we should explicitly state that they can't name a nested struct. What do you think? Should I go for readability or do I preserve the error message?

Reply:
Let's go for simpler code for now, and introduce custom diagnostics if/when requested.

Reply:
Fair enough. Let me bless the tests with the new messages then.